File: tests/e2e/nodejs_client/test/g1_load_extract_avro_orc.test.js
/**
 * E2E: G1 load Avro + extract Avro against a live bqemulator container
 * via the @google-cloud/bigquery Node.js client.
 *
 * The Makefile target (`test-e2e-nodejs`) bind-mounts a host directory
 * onto `/var/lib/bqemu-gcs` and pre-stages canonical Avro/ORC files via
 * `scripts/stage_g1_e2e_fixtures.py`. We reference them through
 * `gs://g1-e2e/<name>` URIs; the executor's `_resolve_uri` strips the
 * scheme and resolves them under `BQEMU_GCS_LOCAL_ROOT`.
 */

const { describe, it, before, after } = require("node:test");
const assert = require("node:assert/strict");

const REST_URL = process.env.BQEMU_REST_URL || "http://localhost:9050";
const PROJECT = "e2e-nodejs-g1";
const DATASET = "g1_node_ds";
const BUCKET = "g1-e2e";

function makeClient() {
  const { BigQuery } = require("@google-cloud/bigquery");
  const { OAuth2Client } = require("google-auth-library");

  const fake = new OAuth2Client();
  fake.credentials = { access_token: "anonymous" };

  return new BigQuery({
    projectId: PROJECT,
    apiEndpoint: REST_URL,
    authClient: fake,
    autoRetry: false,
  });
}

async function postJob(body) {
  const url = `${REST_URL}/bigquery/v2/projects/${PROJECT}/jobs`;
  const res = await fetch(url, {
    method: "POST",
    headers: { "content-type": "application/json" },
    body: JSON.stringify({ configuration: body }),
  });
  if (!res.ok) {
    const text = await res.text();
    throw new Error(`POST ${url} -> ${res.status} ${text}`);
  }
  return res.json();
}

async function cleanup(client) {
  try {
    await client.dataset(DATASET).delete({ force: true });
  } catch (_) {
    /* ignore */
  }
}

describe("bqemulator G1 load/extract Avro+ORC (Node.js)", () => {
  let client;
  before(async () => {
    client = makeClient();
    await client.createDataset(DATASET, { location: "US" }).catch(() => {});
  });
  after(async () => {
    await cleanup(client);
  });

  it("loads a 3-row Avro file via jobs.insert", async () => {
    await client.query(
      `CREATE TABLE IF NOT EXISTS \`${PROJECT}.${DATASET}.items_avro\` ` +
        `(id INT64, name STRING)`,
    );

    await postJob({
      load: {
        destinationTable: {
          projectId: PROJECT,
          datasetId: DATASET,
          tableId: "items_avro",
        },
        sourceUris: [`gs://${BUCKET}/load_avro_basic.avro`],
        sourceFormat: "AVRO",
        writeDisposition: "WRITE_TRUNCATE",
      },
    });

    const [rows] = await client.query(
      `SELECT id, name FROM \`${PROJECT}.${DATASET}.items_avro\` ORDER BY id`,
    );
    assert.equal(rows.length, 3);
    assert.equal(rows[0].name, "alpha");
    assert.equal(rows[2].name, "gamma");
  });

  it("round-trips: extract to Avro, then load it back", async () => {
    // Seed src.
    await client.query(
      `CREATE TABLE IF NOT EXISTS \`${PROJECT}.${DATASET}.rt_src\` ` +
        `(id INT64, val STRING)`,
    );
    await client.query(
      `INSERT INTO \`${PROJECT}.${DATASET}.rt_src\` (id, val) ` +
        `VALUES (1, 'alpha'), (2, 'beta')`,
    );
    await client.query(
      `CREATE TABLE IF NOT EXISTS \`${PROJECT}.${DATASET}.rt_dst\` ` +
        `(id INT64, val STRING)`,
    );

    // Extract.
    await postJob({
      extract: {
        sourceTable: {
          projectId: PROJECT,
          datasetId: DATASET,
          tableId: "rt_src",
        },
        destinationUris: [`gs://${BUCKET}/extract_nodejs.avro`],
        destinationFormat: "AVRO",
      },
    });

    // Load the extract back into rt_dst — proves the Avro file is well-formed
    // without needing a Node Avro decoder.
    await postJob({
      load: {
        destinationTable: {
          projectId: PROJECT,
          datasetId: DATASET,
          tableId: "rt_dst",
        },
        sourceUris: [`gs://${BUCKET}/extract_nodejs.avro`],
        sourceFormat: "AVRO",
        writeDisposition: "WRITE_TRUNCATE",
      },
    });

    const [srcRows] = await client.query(
      `SELECT id, val FROM \`${PROJECT}.${DATASET}.rt_src\` ORDER BY id`,
    );
    const [dstRows] = await client.query(
      `SELECT id, val FROM \`${PROJECT}.${DATASET}.rt_dst\` ORDER BY id`,
    );
    assert.deepEqual(
      srcRows.map((r) => ({ id: Number(r.id), val: r.val })),
      dstRows.map((r) => ({ id: Number(r.id), val: r.val })),
    );
  });
});

File: tests/e2e/go_client/g1_load_extract_avro_orc_test.go
// G1 load Avro + extract Avro E2E against a live bqemulator container.
//
// The Makefile target (`test-e2e-go`) pre-stages canonical Avro/ORC
// fixtures under a bind-mounted host directory; we reference them
// via `gs://g1-e2e/<file>` URIs.

package e2e

import (
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"net/http"
	"testing"
	"time"

	"cloud.google.com/go/bigquery"
	"google.golang.org/api/iterator"
	"google.golang.org/api/option"
)

const (
	g1Project = "e2e-go-g1"
	g1Bucket  = "g1-e2e"
)

func g1Client(ctx context.Context, t *testing.T) *bigquery.Client {
	t.Helper()
	client, err := bigquery.NewClient(
		ctx,
		g1Project,
		option.WithEndpoint(bqAPIBase()),
		option.WithoutAuthentication(),
	)
	if err != nil {
		t.Fatalf("NewClient: %v", err)
	}
	return client
}

func postJob(t *testing.T, configuration map[string]interface{}) {
	t.Helper()
	body, err := json.Marshal(map[string]interface{}{
		"configuration": configuration,
	})
	if err != nil {
		t.Fatalf("marshal: %v", err)
	}
	url := fmt.Sprintf("%s/bigquery/v2/projects/%s/jobs", restURL(), g1Project)
	resp, err := http.Post(url, "application/json", bytes.NewReader(body))
	if err != nil {
		t.Fatalf("POST %s: %v", url, err)
	}
	defer resp.Body.Close()
	if resp.StatusCode < 200 || resp.StatusCode >= 300 {
		var buf bytes.Buffer
		_, _ = buf.ReadFrom(resp.Body)
		t.Fatalf("POST %s -> %d: %s", url, resp.StatusCode, buf.String())
	}
}

func TestG1LoadAvro(t *testing.T) {
	ctx, cancel := context.WithTimeout(context.Background(), 30*time.Second)
	defer cancel()
	client := g1Client(ctx, t)
	defer client.Close()

	ds := client.Dataset("g1_go_avro")
	_ = ds.DeleteWithContents(ctx)
	if err := ds.Create(ctx, &bigquery.DatasetMetadata{Location: "US"}); err != nil {
		t.Fatalf("create dataset: %v", err)
	}
	defer func() { _ = ds.DeleteWithContents(ctx) }()

	tbl := ds.Table("items")
	if err := tbl.Create(ctx, &bigquery.TableMetadata{
		Schema: bigquery.Schema{
			{Name: "id", Type: bigquery.IntegerFieldType, Required: true},
			{Name: "name", Type: bigquery.StringFieldType},
		},
	}); err != nil {
		t.Fatalf("create table: %v", err)
	}

	postJob(t, map[string]interface{}{
		"load": map[string]interface{}{
			"destinationTable": map[string]string{
				"projectId": g1Project,
				"datasetId": "g1_go_avro",
				"tableId":   "items",
			},
			"sourceUris":       []string{fmt.Sprintf("gs://%s/load_avro_basic.avro", g1Bucket)},
			"sourceFormat":     "AVRO",
			"writeDisposition": "WRITE_TRUNCATE",
		},
	})

	q := client.Query("SELECT id, name FROM g1_go_avro.items ORDER BY id")
	it, err := q.Read(ctx)
	if err != nil {
		t.Fatalf("query: %v", err)
	}
	type row struct {
		ID   int64
		Name string
	}
	got := make([]row, 0, 3)
	for {
		var r row
		err := it.Next(&r)
		if err == iterator.Done {
			break
		}
		if err != nil {
			t.Fatalf("iter: %v", err)
		}
		got = append(got, r)
	}
	if len(got) != 3 || got[0].Name != "alpha" || got[2].Name != "gamma" {
		t.Fatalf("unexpected rows: %+v", got)
	}
}

func TestG1ExtractAvroRoundTrip(t *testing.T) {
	ctx, cancel := context.WithTimeout(context.Background(), 30*time.Second)
	defer cancel()
	client := g1Client(ctx, t)
	defer client.Close()

	ds := client.Dataset("g1_go_rt")
	_ = ds.DeleteWithContents(ctx)
	if err := ds.Create(ctx, &bigquery.DatasetMetadata{Location: "US"}); err != nil {
		t.Fatalf("create dataset: %v", err)
	}
	defer func() { _ = ds.DeleteWithContents(ctx) }()

	schema := bigquery.Schema{
		{Name: "id", Type: bigquery.IntegerFieldType, Required: true},
		{Name: "val", Type: bigquery.StringFieldType},
	}
	src := ds.Table("rt_src")
	if err := src.Create(ctx, &bigquery.TableMetadata{Schema: schema}); err != nil {
		t.Fatalf("create src: %v", err)
	}
	dst := ds.Table("rt_dst")
	if err := dst.Create(ctx, &bigquery.TableMetadata{Schema: schema}); err != nil {
		t.Fatalf("create dst: %v", err)
	}

	type srcRow struct {
		ID  int64  `bigquery:"id"`
		Val string `bigquery:"val"`
	}
	if err := src.Inserter().Put(ctx, []srcRow{{1, "x"}, {2, "y"}}); err != nil {
		t.Fatalf("insert: %v", err)
	}

	// Extract.
	postJob(t, map[string]interface{}{
		"extract": map[string]interface{}{
			"sourceTable": map[string]string{
				"projectId": g1Project,
				"datasetId": "g1_go_rt",
				"tableId":   "rt_src",
			},
			"destinationUris":   []string{fmt.Sprintf("gs://%s/extract_go.avro", g1Bucket)},
			"destinationFormat": "AVRO",
		},
	})

	// Re-load (proves the extracted file is valid Avro).
	postJob(t, map[string]interface{}{
		"load": map[string]interface{}{
			"destinationTable": map[string]string{
				"projectId": g1Project,
				"datasetId": "g1_go_rt",
				"tableId":   "rt_dst",
			},
			"sourceUris":       []string{fmt.Sprintf("gs://%s/extract_go.avro", g1Bucket)},
			"sourceFormat":     "AVRO",
			"writeDisposition": "WRITE_TRUNCATE",
		},
	})

	q := client.Query("SELECT id, val FROM g1_go_rt.rt_dst ORDER BY id")
	it, err := q.Read(ctx)
	if err != nil {
		t.Fatalf("query: %v", err)
	}
	type pair struct {
		ID  int64
		Val string
	}
	var got []pair
	for {
		var r pair
		err := it.Next(&r)
		if err == iterator.Done {
			break
		}
		if err != nil {
			t.Fatalf("iter: %v", err)
		}
		got = append(got, r)
	}
	if len(got) != 2 || got[0].Val != "x" || got[1].Val != "y" {
		t.Fatalf("round-trip mismatch: %+v", got)
	}
}
